### ISSUE TITLE
SPIKE: Determine Internal valueset data model

### ISSUE THREAD

**Issue by DanPaseltiner** [CDCgov/phdi#2760]

**Comment by m-goggins** [comment]
### Option #1 (Current)

Here are the current classes we have for valuesets and their items. These are not well named and required storing all of the VS data for each concept, e.g., `ValueSetItem` is actually a concept that perhaps does not need to store the author (this could instead be stored at the ValueSet level). 

```
/*Type to specify the expected components for each item in a value set that will be 
displayed in the CustomizeQuery component*/
export interface ValueSetItem {
  code: string;
  display: string;
  system: string;
  include: boolean;
  author: string;
  clinicalServiceType: string;
  valueSetName: string;
}

/*Type to specify the expected expected types of valueset items that will be displayed 
as separate tabs in the CusomizeQuery component*/
export interface ValueSet {
  labs: ValueSetItem[];
  medications: ValueSetItem[];
  conditions: ValueSetItem[];
}

export type ValueSetType = keyof ValueSet;

export const valueSetTypeToClincalServiceTypeMap = {
  labs: ["ostc", "lotc", "lrtc"],
  medications: ["mrtc"],
  conditions: ["dxtc", "sdtc"],
};

```
### Option #2: Small Modifications

Here's a low lift change that wouldn't require changing any of the current functions:

- Rename `ValueSetItem` to `Concept`
- Add `dibbsClinicalServiceType` to specify if a given concept is a lab, medication, or concept (which would also be in the DB), removing the need for `valueSetTypeToClincalServiceTypeMap`

This moves us towards a better framework of a valueset without requiring us to re-work the logic to display customized queries. However, like Option #1, it doesn't store all of the data at the correct level

```
export interface Concept{
  code: string;
  display: string;
  system: string;
  include: boolean;
  author: string;
  clinicalServiceType: string; // this is lrtc, mrtc, etc.
  dibbsClinicalServiceType: string; // this would be labs/medications/conditions 
  valueSetName: string;
}

export interface ValueSet {
  labs: Concept[];
  medications: Concept[];
  conditions: Concept[];
}
```

### Option #3: Rework
Here's a more accurate representation of what a valueset actually "looks" like in the DB and correctly stores all of the data at the appropriate level, i.e., concept data is in the concept type and valueset data is in the valueset type. This change would require some additional work to display the data correctly on the CustomQuery page and possibly on running the queries themselves.

Two questions:

- Should we keep track of "include" at both the concept and valueset levels? Currently we keep track only at the concept level and use `handleSelectAllChange` to check/uncheck every concept in a valueset. I'm wondering if it is also worth it to track 'includeValueSet`, though we would still have to check/uncheck every concept.
- Should we track version and valueset_id in this internal data model? I think we need this in order to check for updates and potentially to display the changes to users in the future when we figure out how to show differences, e.g., we need to know which vs is version 2 and which is version 3 in order to display the differences.

```
export interface Concept {
  code: string;
  display: string;
  include: boolean;
}

export interface ValueSet {
 valueset_id: string, // this will help us track which version users have stored
 valueset_version: string,
 valueset_name: string;
 author: string;
 clinicalServiceType: string; // this is lrtc, mrtc, etc.
 dibbsClinicalServiceType: string; // this would be labs/medications/conditions
 includeValueSet: boolean; **NOT SURE ABOUT THIS**
 concepts: Concept[]
}

export interface ValueSetDisplay {
 labs: ValueSet[],
 medications: ValueSet[],
 conditions: Valueset[]
}
```

I'm leaning towards Option #3 because I think it is most accurate to the correct data model, even if it requires more time to implement it correctly. 

### Implementation

- Users can select/deselect individual codes
- Users can select/deselect whole value sets
- Users can add individual codes 
- Updates to value sets can be handled

**Comment by DanPaseltiner** [comment]
I'm leaning towards option 3 as well. However, I don't see any place to store `system`. For both of the service type keys we should define an enum, also maybe we should change the names of these fields to `dibbsConceptType` and `ersdConceptType`? I think I initially introduced the language of "clinical service" and that may have been a mistake.

**Comment by m-goggins** [comment]
Yes, updated to address all of your comments. I like `ConceptType` because it more closely matches what's in the db
```
export interface Concept {
  code: string;
  display: string;
  include: boolean;
}

export interface ValueSet {
 valueset_id: string, // this will help us track which version users have stored
 valueset_version: string,
 valueset_name: string;
 author: string;
 system: string;
 ersdConceptType?: string; // this is lrtc, mrtc, etc.
 dibbsConceptType: string; // this would be labs/medications/conditions
 includeValueSet: boolean;
 concepts: Concept[]
}

export interface ValueSetDisplay {
 labs: ValueSet[],
 medications: ValueSet[],
 conditions: Valueset[]
}
```
